Session Lifespan Feature

hoseinrafiei opened the pull request:
Added a new feature to manage session lifespan in **Registration Settings**.
Fixes https://github.com/eventespresso/event-espresso-core/issues/3871 issue.

Added a test class to test the functionality of Session Lifespan Option.
Updated the WP GraphQL package.

hoseinrafiei commented:
@tn3rb

Please check this. All the items that you described here are done: https://github.com/eventespresso/event-espresso-core/issues/3871
I also updated the `wp-graphql` package to the latest production version and the tests are running fine now but we may check the graphql API to make sure there is no problem with the latest package.

Also the test class is added and works fine. Your feedback on the test class is highly appreciated.

tn3rb commented on caffeinated/admin/extend/registration_form/Extend_Registration_Form_Admin_Page.core.php:
```diff
@@ -1247,15 +1255,13 @@ public function update_copy_attendee_info_settings_form(EE_Registration_Config $
                         );
                     }
-                } else {
-                    if ($copy_attendee_info_settings_form->submission_error_message() !== '') {
-                        EE_Error::add_error(
-                            $copy_attendee_info_settings_form->submission_error_message(),
-                            __FILE__,
-                            __FUNCTION__,
-                            __LINE__
-                        );
-                    }
+                } elseif ($copy_attendee_info_settings_form->submission_error_message() !== '') {
```
👍🏻 good stuff

tn3rb reviewed (changes requested):
This all looks really really good...

I did have a couple of requests for a few things because I changed my mind about how to approach things (not because of you, you're code was great)

tn3rb commented:
just looked, and you can ignore this failed unit test:

```
1) CachingLoaderTest::testResetCache
Failed asserting that actual size 3 matches expected size 0.
```

it does not happen locally and I am currently trying to figure it out

hoseinrafiei commented on caffeinated/admin/extend/registration_form/forms/SessionLifespanForm.php:
```diff
@@ -0,0 +1,62 @@
+     */
+    public function __construct()
+    {
+        $this->sessionLifespanOption = new SessionLifespanOption();
```
Done.

hoseinrafiei commented on caffeinated/admin/extend/registration_form/forms/SessionLifespanFormHandler.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * @var SessionLifespanOption
+     */
+    private $sessionLifespanOption;
```
👍

hoseinrafiei commented on caffeinated/admin/extend/registration_form/forms/SessionLifespanFormHandler.php:
```diff
@@ -0,0 +1,70 @@
+    public function __construct()
+    {
+        $this->sessionLifespanOption = new SessionLifespanOption();
+    }
```
👍

hoseinrafiei commented on core/domain/values/session/SessionLifespan.php:
```diff
@@ -37,7 +37,10 @@ class SessionLifespan
+            $lifespan = $lifespanOption->getSessionLifespan();
+        }
         $this->setLifespan($lifespan);
     }
```
Yes, It was good idea. Thanks for sharing. I did it.

hoseinrafiei commented:
@tn3rb Please check it again. I implemented your requested changes.

tn3rb commented on core/EE_Dependency_Map.core.php:
```diff
@@ -578,6 +578,9 @@ protected function _register_core_dependencies()
             'EventEspresso\core\domain\services\commands\attendee\CreateAttendeeCommandHandler'                                  => [
                 'EEM_Attendee' => EE_Dependency_Map::load_from_cache,
             ],
+            'EventEspresso\core\domain\values\session\SessionLifespan'                                                    => [
+                'EventEspresso\core\domain\values\session\SessionLifespanOption' => EE_Dependency_Map::load_from_cache,
+            ],
```
### AWESOME !!! 😀 

wanted to see if you could figure that out yourself!

tn3rb commented on core/domain/values/session/SessionLifespan.php:
```diff
@@ -31,17 +31,12 @@ class SessionLifespan
-            $lifespan = $lifespanOption->getSessionLifespan();
-        }
-        $this->setLifespan($lifespan);
+        $this->setLifespan($session_lifespan_option->getSessionLifespan());
```
👍🏻

tn3rb commented on caffeinated/admin/extend/registration_form/Extend_Registration_Form_Admin_Page.core.php:
```diff
@@ -1462,14 +1463,15 @@ private function _verify_pcre_support(EE_Registration_Config $EE_Registration_Co
     public function setSessionLifespan()
     {
-        $session_lifespan_form = new SessionLifespanForm();
+        $session_lifespan_form = new SessionLifespanForm(new SessionLifespanOption());
```
if we use the loader, then any dependencies can be automatically injected

```php
$session_lifespan_form = $this->loader->getNew(SessionLifespanForm::class);
```

`$this->loader` should be available within all admin pages

hoseinrafiei replied:
Done.

hoseinrafiei commented:
@tn3rb Please check it again.

tn3rb reviewed (approved):
### Great Job Hossein 
# 💯 ✖️ 🌮 ➡️ 🦖

tn3rb commented:
@hoseinrafiei 

Once final request... I'm concerned by the 900,000 lines of code being removed from the codebase as part of this PR.
Was there a specific reason why WP GQL needed to be updated as part of this PR?
I'm thinking that that update should be split out into it's own PR because if it causes problems then we may need to revert it, which we won't be able to do if it is embedded within this PR.

hoseinrafiei commented:
> @hoseinrafiei
> 
> Once final request... I'm concerned by the 900,000 lines of code being removed from the codebase as part of this PR. Was there a specific reason why WP GQL needed to be updated as part of this PR? I'm thinking that that update should be split out into it's own PR because if it causes problems then we may need to revert it, which we won't be able to do if it is embedded within this PR.

@tn3rb

It was to fix the issue of phpunit tests and writing proper test for the new class.
We discussed in Slack channel about it.

I'm fine to remove it and create a separate PR for it. Just let me know how to do it. Just point me to the right direction and then I will search and find.

hoseinrafiei commented:
@tn3rb I moved the WP GraphQL update to https://github.com/eventespresso/event-espresso-core/pull/3993 branch and removed it from this branch. Please check it and let me know if anything else is needed.

tn3rb commented:
ok this is looking great, just have a few things that need to be cleaned up before we can merge this:

- there's some linting errors
- needs to be rebased
- the `composer.json` file needs to be updated 


#### `composer.json` update

- add top level item named `allow-plugins`
- add keys for each package listed in `require` or `require-dev`
- set values to `true`
- look at EE core for an example

tn3rb reviewed (changes requested):
see above note

hoseinrafiei commented:
@tn3rb Please check it again. The composer.json file seems to be fixed automatically after rebase.

tn3rb reviewed (approved):
# Great Job 💯

garthkoyle commented:
@hoseinrafiei I don't see this working. I have registrations that are "incomplete" status (no registration form info submitted), and their status is still Incomplete way after the time period selected in the Registration Settings.

Currently, my session length is set to 5 minutes: 
![image](https://user-images.githubusercontent.com/3094547/188992432-2d00270e-f1c8-4d9d-870c-b37e388e57e9.png)

The lifespan session length of time has passed, and at least the Incomplete registrations are still `Incomplete`. Based on Brent's description, I expected the `Incomplete` registrations to be set to `Canceled` and thus do not count against the capacity limits set for the datetime and/or ticket.

![image](https://user-images.githubusercontent.com/3094547/188992971-33403b4e-f787-49ac-b656-f54d7c7bcf78.png)


Does the session lifespan apply to each registration individually or does the lifespan time period apply to all registrations no matter when received? For example, if my registration lifespan period was set to 1 hour when I registered 20 minutes ago, and now I set my lifespan to 5 minutes and then register again, what lifespan time period will be applied to each of those two registrations?

Does this code only apply to "Incomplete" registrations or all registration types, e.g. Pending Payment, etc?

garthkoyle commented:
With more testing I discovered about the lifespan feature. When the lifespan feature triggers according to the selected time period:

- Free or paid registrations without any registration information (Incomplete Registrations), the registration status is never changed in the admin. Only the event tickets become available after the hold time period expires.
- Paid registrations WITH registration form information (Pending Payment) are NOT ever changed in the admin, but the tickets become available after the hold time expires.

So, this probably functions correctly, but maybe the only change I might recommend is to set the registration status to `Canceled` for `Incomplete` registrations only. Would that make sense?

hoseinrafiei commented:
@garthkoyle 

Thanks for your feedback. This task was to add an ability to change the **session lifespan time** using a configuration form and some code refactor. Previously the time could change only through the code using a WordPress filter.

Changing the status was not included in the task description, but it looks great to add this ability in my idea.
I'm just not sure if `Canceled` is a suitable term for `Incomplete` registrations since we have to differentiate between those registrations that really canceled by user (after complete booking) and those registrations that are `Abandoned` during the booking process.

What do you think?

hoseinrafiei commented:
@tn3rb I think it will be good if you get involved and provide your feedback and also let me know what is the best way to implement the change status function?
